File: surface/widgets/compose/components/widgets/concepts/formal-verification/TraceTimelineViewer.kt
package com.clef.surface.widgets.concepts.formalverification

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.rememberScrollState
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.delay

// --- State machine ---

enum class TraceTimelineViewerState { Idle, Playing, CellSelected }

sealed class TraceTimelineViewerEvent {
    object Play : TraceTimelineViewerEvent()
    object StepForward : TraceTimelineViewerEvent()
    object StepBackward : TraceTimelineViewerEvent()
    object SelectCell : TraceTimelineViewerEvent()
    object Zoom : TraceTimelineViewerEvent()
    object Pause : TraceTimelineViewerEvent()
    object StepEnd : TraceTimelineViewerEvent()
    object Deselect : TraceTimelineViewerEvent()
}

fun traceTimelineViewerReduce(state: TraceTimelineViewerState, event: TraceTimelineViewerEvent): TraceTimelineViewerState = when (state) {
    TraceTimelineViewerState.Idle -> when (event) {
        is TraceTimelineViewerEvent.Play -> TraceTimelineViewerState.Playing
        is TraceTimelineViewerEvent.StepForward -> TraceTimelineViewerState.Idle
        is TraceTimelineViewerEvent.StepBackward -> TraceTimelineViewerState.Idle
        is TraceTimelineViewerEvent.SelectCell -> TraceTimelineViewerState.CellSelected
        is TraceTimelineViewerEvent.Zoom -> TraceTimelineViewerState.Idle
        else -> state
    }
    TraceTimelineViewerState.Playing -> when (event) {
        is TraceTimelineViewerEvent.Pause -> TraceTimelineViewerState.Idle
        is TraceTimelineViewerEvent.StepEnd -> TraceTimelineViewerState.Idle
        else -> state
    }
    TraceTimelineViewerState.CellSelected -> when (event) {
        is TraceTimelineViewerEvent.Deselect -> TraceTimelineViewerState.Idle
        is TraceTimelineViewerEvent.SelectCell -> TraceTimelineViewerState.CellSelected
        else -> state
    }
}

// --- Types ---

data class TraceStep(
    val index: Int,
    val label: String,
    val state: Map<String, String>,
    val isError: Boolean = false,
    val timestamp: String? = null
)

data class SelectedCell(val step: Int, val variable: String)

@Composable
fun TraceTimelineViewer(
    steps: List<TraceStep>,
    modifier: Modifier = Modifier,
    variables: List<String>? = null,
    currentStep: Int? = null,
    playbackSpeed: Float = 1.0f,
    showChangesOnly: Boolean = false,
    zoom: Float = 1.0f,
    onStepChange: (Int) -> Unit = {}
) {
    var widgetState by remember { mutableStateOf(TraceTimelineViewerState.Idle) }
    var internalStep by remember { mutableIntStateOf(0) }
    val activeStep = currentStep ?: internalStep
    var selectedCell by remember { mutableStateOf<SelectedCell?>(null) }
    var focusedLane by remember { mutableIntStateOf(0) }

    // Derive variable names from steps if not provided
    val resolvedVariables = remember(variables, steps) {
        variables ?: steps.flatMap { it.state.keys }.distinct()
    }

    // Sync with controlled prop
    LaunchedEffect(currentStep) {
        if (currentStep != null) internalStep = currentStep
    }

    fun goToStep(idx: Int) {
        val clamped = idx.coerceIn(0, (steps.size - 1).coerceAtLeast(0))
        internalStep = clamped
        onStepChange(clamped)
    }

    fun didValueChange(stepIdx: Int, variable: String): Boolean {
        if (stepIdx == 0) return false
        val prev = steps.getOrNull(stepIdx - 1)?.state?.get(variable)
        val curr = steps.getOrNull(stepIdx)?.state?.get(variable)
        return prev != curr
    }

    // Playback timer
    LaunchedEffect(widgetState, playbackSpeed) {
        if (widgetState == TraceTimelineViewerState.Playing) {
            val intervalMs = maxOf(100L, (1000L / playbackSpeed).toLong())
            while (true) {
                delay(intervalMs)
                val next = internalStep + 1
                if (next >= steps.size) {
                    widgetState = traceTimelineViewerReduce(widgetState, TraceTimelineViewerEvent.StepEnd)
                    break
                }
                internalStep = next
                onStepChange(next)
            }
        }
    }

    val currentStepData = steps.getOrNull(activeStep)
    val scrollState = rememberScrollState()

    Column(modifier = modifier.semantics { contentDescription = "Trace timeline" }) {
        // Time axis
        Row(
            modifier = Modifier.fillMaxWidth().horizontalScroll(scrollState).padding(start = 80.dp),
            horizontalArrangement = Arrangement.spacedBy(4.dp)
        ) {
            steps.forEach { step ->
                Text(
                    "${step.index}",
                    fontSize = 11.sp,
                    fontFamily = FontFamily.Monospace,
                    color = if (step.isError) MaterialTheme.colorScheme.error else MaterialTheme.colorScheme.onSurfaceVariant,
                    fontWeight = if (step.index == activeStep) FontWeight.Bold else FontWeight.Normal,
                    modifier = Modifier.width((40 * zoom).dp)
                )
            }
        }

        HorizontalDivider()

        // Variable lanes
        LazyColumn(modifier = Modifier.weight(1f)) {
            itemsIndexed(resolvedVariables) { laneIdx, variable ->
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    modifier = Modifier
                        .fillMaxWidth()
                        .then(if (laneIdx == focusedLane) Modifier.background(MaterialTheme.colorScheme.surfaceVariant) else Modifier)
                ) {
                    // Lane label
                    Text(
                        variable,
                        fontSize = 12.sp,
                        fontFamily = FontFamily.Monospace,
                        modifier = Modifier.width(80.dp).padding(horizontal = 4.dp)
                    )

                    // Cells
                    Row(
                        modifier = Modifier.horizontalScroll(scrollState),
                        horizontalArrangement = Arrangement.spacedBy(4.dp)
                    ) {
                        steps.forEach { step ->
                            val value = step.state[variable] ?: ""
                            val changed = didValueChange(step.index, variable)
                            if (showChangesOnly && !changed && step.index != 0) return@forEach

                            val isCurrent = step.index == activeStep
                            val isSelected = selectedCell?.step == step.index && selectedCell?.variable == variable

                            Surface(
                                tonalElevation = if (isCurrent) 2.dp else 0.dp,
                                modifier = Modifier
                                    .width((40 * zoom).dp)
                                    .clickable {
                                        selectedCell = SelectedCell(step.index, variable)
                                        goToStep(step.index)
                                        widgetState = traceTimelineViewerReduce(widgetState, TraceTimelineViewerEvent.SelectCell)
                                    }
                                    .then(
                                        if (isSelected) Modifier.border(2.dp, MaterialTheme.colorScheme.primary)
                                        else Modifier
                                    )
                                    .then(
                                        if (step.isError) Modifier.background(Color(0x20EF4444))
                                        else Modifier
                                    )
                            ) {
                                Text(
                                    value,
                                    fontSize = 11.sp,
                                    fontFamily = FontFamily.Monospace,
                                    fontWeight = if (changed) FontWeight.Bold else FontWeight.Normal,
                                    color = if (step.isError) MaterialTheme.colorScheme.error else MaterialTheme.colorScheme.onSurface,
                                    modifier = Modifier.padding(2.dp),
                                    maxLines = 1
                                )
                            }
                        }
                    }
                }
            }
        }

        HorizontalDivider()

        // Playback controls
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(4.dp),
            modifier = Modifier.padding(8.dp)
        ) {
            IconButton(
                onClick = { widgetState = traceTimelineViewerReduce(widgetState, TraceTimelineViewerEvent.StepBackward); goToStep(activeStep - 1) },
                enabled = activeStep > 0
            ) { Text("\u00AB", fontSize = 16.sp) }

            IconButton(onClick = {
                widgetState = if (widgetState == TraceTimelineViewerState.Playing) {
                    traceTimelineViewerReduce(widgetState, TraceTimelineViewerEvent.Pause)
                } else {
                    traceTimelineViewerReduce(widgetState, TraceTimelineViewerEvent.Play)
                }
            }) {
                Text(if (widgetState == TraceTimelineViewerState.Playing) "\u23F8" else "\u25B6", fontSize = 16.sp)
            }

            IconButton(
                onClick = { widgetState = traceTimelineViewerReduce(widgetState, TraceTimelineViewerEvent.StepForward); goToStep(activeStep + 1) },
                enabled = activeStep < steps.size - 1
            ) { Text("\u00BB", fontSize = 16.sp) }

            Text(
                if (steps.isNotEmpty()) "${activeStep + 1} / ${steps.size}" else "0 / 0",
                fontSize = 13.sp
            )
        }

        // Detail panel when a cell is selected
        if (widgetState == TraceTimelineViewerState.CellSelected && currentStepData != null) {
            HorizontalDivider()
            Column(Modifier.padding(12.dp)) {
                Text(
                    "Step ${currentStepData.index}: ${currentStepData.label}" +
                        if (currentStepData.isError) " (error)" else "",
                    fontWeight = FontWeight.Bold,
                    color = if (currentStepData.isError) MaterialTheme.colorScheme.error else MaterialTheme.colorScheme.onSurface
                )
                currentStepData.timestamp?.let { Text(it, fontSize = 12.sp, color = MaterialTheme.colorScheme.onSurfaceVariant) }
                Spacer(Modifier.height(4.dp))
                currentStepData.state.forEach { (key, value) ->
                    val changed = didValueChange(activeStep, key)
                    Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                        Text(key, fontSize = 12.sp, fontFamily = FontFamily.Monospace)
                        Text(
                            value,
                            fontSize = 12.sp,
                            fontFamily = FontFamily.Monospace,
                            fontWeight = if (changed) FontWeight.Bold else FontWeight.Normal
                        )
                    }
                }
            }
        }
    }
}

File: surface/widgets/nativescript/components/widgets/concepts/formal-verification/TraceTimelineViewer.ts
import {
  StackLayout,
  GridLayout,
  Label,
  Button,
  ScrollView,
  Color,
  View,
} from '@nativescript/core';

export type TraceTimelineViewerState = 'idle' | 'playing' | 'cellSelected';
export type TraceTimelineViewerEvent =
  | { type: 'PLAY' }
  | { type: 'STEP_FORWARD' }
  | { type: 'STEP_BACKWARD' }
  | { type: 'SELECT_CELL' }
  | { type: 'ZOOM' }
  | { type: 'PAUSE' }
  | { type: 'STEP_END' }
  | { type: 'DESELECT' };

export function traceTimelineViewerReducer(state: TraceTimelineViewerState, event: TraceTimelineViewerEvent): TraceTimelineViewerState {
  switch (state) {
    case 'idle':
      if (event.type === 'PLAY') return 'playing';
      if (event.type === 'STEP_FORWARD') return 'idle';
      if (event.type === 'STEP_BACKWARD') return 'idle';
      if (event.type === 'SELECT_CELL') return 'cellSelected';
      if (event.type === 'ZOOM') return 'idle';
      return state;
    case 'playing':
      if (event.type === 'PAUSE') return 'idle';
      if (event.type === 'STEP_END') return 'idle';
      return state;
    case 'cellSelected':
      if (event.type === 'DESELECT') return 'idle';
      if (event.type === 'SELECT_CELL') return 'cellSelected';
      return state;
    default:
      return state;
  }
}

export interface TraceStep {
  index: number;
  label: string;
  state: Record<string, string>;
  isError?: boolean;
  timestamp?: string;
}

export interface TraceTimelineViewerProps {
  steps: TraceStep[];
  variables?: string[];
  currentStep?: number;
  playbackSpeed?: number;
  showChangesOnly?: boolean;
  zoom?: number;
  onStepChange?: (stepIndex: number) => void;
}

export function createTraceTimelineViewer(props: TraceTimelineViewerProps): { view: View; dispose: () => void } {
  let widgetState: TraceTimelineViewerState = 'idle';
  let activeStep = props.currentStep ?? 0;
  let selectedCell: { step: number; variable: string } | null = null;
  let playbackTimer: ReturnType<typeof setInterval> | null = null;
  const disposers: (() => void)[] = [];

  const variables: string[] = props.variables ?? (() => {
    const keys = new Set<string>();
    for (const step of props.steps) for (const k of Object.keys(step.state)) keys.add(k);
    return Array.from(keys);
  })();

  function send(event: TraceTimelineViewerEvent) {
    widgetState = traceTimelineViewerReducer(widgetState, event);
    render();
  }

  function goToStep(idx: number) {
    activeStep = Math.max(0, Math.min(idx, props.steps.length - 1));
    props.onStepChange?.(activeStep);
  }

  function didChange(stepIdx: number, variable: string): boolean {
    if (stepIdx === 0) return false;
    return props.steps[stepIdx - 1]?.state[variable] !== props.steps[stepIdx]?.state[variable];
  }

  const root = new StackLayout();
  root.className = 'clef-trace-timeline-viewer';
  root.automationText = 'Trace timeline';

  function render() {
    root.removeChildren();

    // Time axis
    const timeRow = new ScrollView();
    timeRow.orientation = 'horizontal';
    const timeBar = new StackLayout();
    timeBar.orientation = 'horizontal';
    timeBar.padding = '4';
    const corner = new Label();
    corner.text = '';
    corner.width = 80;
    timeBar.addChild(corner);
    for (const step of props.steps) {
      const lbl = new Label();
      lbl.text = String(step.index);
      lbl.width = 60;
      lbl.textAlignment = 'center';
      lbl.fontSize = 11;
      if (step.isError) lbl.color = new Color('#ef4444');
      if (step.index === activeStep) lbl.fontWeight = 'bold';
      timeBar.addChild(lbl);
    }
    timeRow.content = timeBar;
    root.addChild(timeRow);

    // Variable lanes
    const scroll = new ScrollView();
    const lanes = new StackLayout();
    variables.forEach((variable) => {
      const lane = new ScrollView();
      lane.orientation = 'horizontal';
      const row = new StackLayout();
      row.orientation = 'horizontal';
      row.padding = '2 0';

      const varLabel = new Label();
      varLabel.text = variable;
      varLabel.width = 80;
      varLabel.fontSize = 12;
      varLabel.fontWeight = 'bold';
      row.addChild(varLabel);

      for (const step of props.steps) {
        const value = step.state[variable] ?? '';
        const changed = didChange(step.index, variable);
        if (props.showChangesOnly && !changed && step.index !== 0) {
          const empty = new Label();
          empty.text = '';
          empty.width = 60;
          row.addChild(empty);
          continue;
        }
        const cell = new Label();
        cell.text = value;
        cell.width = 60;
        cell.textAlignment = 'center';
        cell.fontSize = 11;
        if (changed) cell.fontWeight = 'bold';
        if (step.isError) cell.backgroundColor = new Color('#fee2e2');
        if (step.index === activeStep) cell.backgroundColor = new Color('#dbeafe');
        if (selectedCell?.step === step.index && selectedCell?.variable === variable) {
          cell.borderWidth = 2;
          cell.borderColor = new Color('#6366f1');
        }
        const cellHandler = () => {
          selectedCell = { step: step.index, variable };
          goToStep(step.index);
          send({ type: 'SELECT_CELL' });
        };
        cell.on('tap', cellHandler);
        disposers.push(() => cell.off('tap', cellHandler));
        row.addChild(cell);
      }
      lane.content = row;
      lanes.addChild(lane);
    });
    scroll.content = lanes;
    root.addChild(scroll);

    // Controls
    const controls = new StackLayout();
    controls.orientation = 'horizontal';
    controls.horizontalAlignment = 'center';
    controls.padding = '8';

    const backBtn = new Button();
    backBtn.text = '\u00AB';
    backBtn.isEnabled = activeStep > 0;
    const bh = () => { send({ type: 'STEP_BACKWARD' }); goToStep(activeStep - 1); };
    backBtn.on('tap', bh);
    disposers.push(() => backBtn.off('tap', bh));
    controls.addChild(backBtn);

    const ppBtn = new Button();
    ppBtn.text = widgetState === 'playing' ? '\u23F8' : '\u25B6';
    const pph = () => {
      if (widgetState === 'playing') send({ type: 'PAUSE' });
      else send({ type: 'PLAY' });
    };
    ppBtn.on('tap', pph);
    disposers.push(() => ppBtn.off('tap', pph));
    controls.addChild(ppBtn);

    const fwdBtn = new Button();
    fwdBtn.text = '\u00BB';
    fwdBtn.isEnabled = activeStep < props.steps.length - 1;
    const fh = () => { send({ type: 'STEP_FORWARD' }); goToStep(activeStep + 1); };
    fwdBtn.on('tap', fh);
    disposers.push(() => fwdBtn.off('tap', fh));
    controls.addChild(fwdBtn);

    const stepLabel = new Label();
    stepLabel.text = props.steps.length > 0 ? ` ${activeStep + 1} / ${props.steps.length}` : ' 0 / 0';
    stepLabel.fontSize = 13;
    stepLabel.marginLeft = 8;
    controls.addChild(stepLabel);

    root.addChild(controls);

    // Playback timer
    if (widgetState === 'playing' && !playbackTimer) {
      const ms = Math.max(100, (1 / (props.playbackSpeed ?? 1)) * 1000);
      playbackTimer = setInterval(() => {
        if (activeStep >= props.steps.length - 1) { send({ type: 'STEP_END' }); return; }
        goToStep(activeStep + 1);
        render();
      }, ms);
      disposers.push(() => { if (playbackTimer) { clearInterval(playbackTimer); playbackTimer = null; } });
    } else if (widgetState !== 'playing' && playbackTimer) {
      clearInterval(playbackTimer);
      playbackTimer = null;
    }

    // Detail panel
    if (widgetState === 'cellSelected' && props.steps[activeStep]) {
      const detail = new StackLayout();
      detail.padding = '8 12';
      detail.borderTopWidth = 1;
      detail.borderTopColor = new Color('#e5e7eb');
      const step = props.steps[activeStep];
      const title = new Label();
      title.text = `Step ${step.index}: ${step.label}${step.isError ? ' (error)' : ''}`;
      title.fontWeight = 'bold';
      title.fontSize = 14;
      detail.addChild(title);
      if (step.timestamp) {
        const ts = new Label();
        ts.text = step.timestamp;
        ts.fontSize = 12;
        ts.color = new Color('#6b7280');
        detail.addChild(ts);
      }
      for (const [k, v] of Object.entries(step.state)) {
        const entry = new Label();
        const ch = didChange(activeStep, k);
        entry.text = `${k}: ${v}`;
        entry.fontSize = 13;
        if (ch) entry.fontWeight = 'bold';
        detail.addChild(entry);
      }
      root.addChild(detail);
    }
  }

  render();

  return {
    view: root,
    dispose() { disposers.forEach(d => d()); if (playbackTimer) clearInterval(playbackTimer); },
  };
}

export default createTraceTimelineViewer;
